### PR TITLE
Simplify HashJoin memory accounting

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change

The use of shared memory reservations was not only confusing, but also unnecessary. This reworks the memory accounting to not need this complexity. If we like this approach I will do something similar for the other joins

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 1]
```diff
@@ -113,8 +109,6 @@ pub struct HashJoinExec {
     schema: SchemaRef,
     /// Build-side data
     left_fut: OnceAsync<JoinLeftData>,
-    /// Operator-level memory reservation for left data
-    reservation: SharedOptionalMemoryReservation,
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:117]: This reservation is instead moved into JoinLeftData

File: datafusion/core/src/physical_plan/joins/hash_join.rs [hunk 2]
```diff
@@ -362,45 +355,35 @@ impl ExecutionPlan for HashJoinExec {
...
-        // Memory reservation for left-side data depends on PartitionMode:
-        // - operator-level for `CollectLeft` mode
-        // - stream-level for partitioned mode
-        //
-        // This approach allows to avoid cases when left data could potentially
-        // outlive its memory reservation and rely on `MemoryReservation` destructors
-        // for releasing memory in pool.
         let left_fut = match self.mode {
             PartitionMode::CollectLeft => self.left_fut.once(|| {
+                let reservation =
```
Comment (alamb) [inline datafusion/core/src/physical_plan/joins/hash_join.rs:360]: I wonder if the same simplification can be applied to Nested Loops Joins and Symmetric Hash Join 🤔 

https://github.com/search?q=repo%3Aapache%2Farrow-datafusion%20SharedMemoryReservation&type=code

**Review by alamb — COMMENTED** [review]
@korowa I wonder if you have some time to review this PR (as I think you have worked on the join memory reservation code most recently)?

**Comment by korowa** [comment]
LGTM, thank you @tustvold!

This version looks way cleaner, indeed.

**Pull request merged**